Internal clients unable to set context menu images

lilyspiniolas opened the pull request:
#### c6bfe500c5dc35b74d322616cb4da3807aedad7a
<pre>
Internal clients unable to set context menu images
<a href="https://bugs.webkit.org/show_bug.cgi?id=285317">https://bugs.webkit.org/show_bug.cgi?id=285317</a>
<a href="https://rdar.apple.com/139537720">rdar://139537720</a>

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

Allow internal WebKit clients to enable context menu images.
When enabled, the ContextMenuAction of menu items is used
to determine which image to use, if available.

PDFContextMenuItems now store a ContextMenuAction so that
the image can be determined in the same manner. Each item&apos;s
ContextMenuAction is determined from its ContextMenuItemTag.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/PAL/pal/spi/mac/NSMenuSPI.h:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKit/Platform/mac/MenuUtilities.h:
* Source/WebKit/Platform/mac/MenuUtilities.mm:
(WebKit::symbolNameForAction):
* Source/WebKit/Shared/mac/PDFContextMenu.h:
* Source/WebKit/Shared/mac/PDFContextMenuItem.serialization.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::platformMenuItems const):
(WebKit::WebExtensionContext::singleMenuItemOrExtensionItemWithSubmenu const):
(WebKit::WebExtensionContext::addItemsToContextMenu):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::updateMenuItemImage):
(WebKit::WebContextMenuProxyMac::getContextMenuItem):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::showPDFContextMenu):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::handleContextMenuEvent):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::contextMenuActionFromTag const):
(WebKit::UnifiedPDFPlugin::contextMenuItem const):
(WebKit::UnifiedPDFPlugin::separatorContextMenuItem const):

Canonical link: <a href="https://commits.webkit.org/288471@main">https://commits.webkit.org/288471@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c17a19b0e6895ce4c9caa8fde3ff578a8ae8d15

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83463 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3080 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88539 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34473 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85551 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3165 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11038 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64931 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22673 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86513 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2314 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75841 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45222 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2219 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30047 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33523 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/76426 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73313 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30783 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89915 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/82482 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10728 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7735 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73356 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10953 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71669 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72585 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17964 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16819 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15537 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2061 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10683 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16152 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104902 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10534 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25361 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14005 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12305 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/PAL/pal/spi/mac/NSMenuSPI.h:
```diff
@@ -67,10 +72,17 @@ enum {
 @class QLPreviewMenuItem;
 
-@interface NSMenuItem ()
+@interface NSMenuItem (Staging_139537720)
```
This would imply the entire category is for staging related to rdar://139537720, but that's not the case.

aprotyas commented on Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
```diff
@@ -884,9 +924,15 @@ - (void)menuDidClose:(NSMenu *)menu
-    case WebCore::ContextMenuItemType::CheckableAction:
-        completionHandler(createMenuActionItem(item).get());
+    case WebCore::ContextMenuItemType::CheckableAction: {
+        auto menuItem = createMenuActionItem(item);
```
```suggestion
        RetainPtr menuItem = createMenuActionItem(item);
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2258,6 +2258,44 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
+        return ContextMenuItemPDFZoomOut;
+    default:
+        return ContextMenuItemTagNoAction;
+    }
```
Minor nit: I would drop the `default` case and return `ContextMenuItemTagNoAction` outside of the `switch`. That way the compiler lets us know when we unknowingly don't handle a `ContextMenuItemTag` case (say if we add a new value).

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1128,7 +1128,7 @@ static bool getEventTypeFromWebEvent(const WebEvent& event, NSEventType& eventTy
         if ([NSStringFromSelector(item.action) isEqualToString:@"openWithPreview"])
             openInPreviewTag = i;
-        PDFContextMenuItem menuItem { String([item title]), static_cast<int>([item state]), i,
+        PDFContextMenuItem menuItem { String([item title]), static_cast<int>([item state]), i, static_cast<ContextMenuAction>([item tag]),
```
Oh TIL: `NSMenuItem.tag` maps exactly to the values in `ContextMenuAction`?

pxlcoder commented on Source/WebKit/Platform/mac/MenuUtilities.mm:
```diff
@@ -33,6 +33,7 @@
 #import <WebCore/RevealUtilities.h>
 
 #if ENABLE(TELEPHONE_NUMBER_DETECTION)
+#import <pal/spi/mac/NSMenuSPI.h>
```
I don't think this needs to be under `ENABLE(TELEPHONE_NUMBER_DETECTION)`.

pxlcoder commented on Source/WebKit/Platform/mac/MenuUtilities.mm:
```diff
@@ -153,6 +154,171 @@ - (BOOL)revealContext:(RVPresentingContext *)context shouldUseDefaultHighlightFo
+    case WebCore::ContextMenuItemTagUnderline:
+        return @"underline";
+    default:
+        return nil;
```
Echoing Abrar's comment about avoiding the use of `default`.

whsieh commented on Source/WebCore/PAL/pal/spi/mac/NSMenuSPI.h:
```diff
@@ -35,6 +35,11 @@
 
+#if ENABLE(CONTEXT_MENU_IMAGES_FOR_INTERNAL_CLIENTS)
++ (NSString *)_systemImageNameForAction:(SEL)action;
+@property (strong, setter=_setActionImage:) NSImage *_actionImage;
```
I think these probably need to go into their own staging category? Rather than being related to rdar://129192954 (WritingTools)

```
#if ENABLE(CONTEXT_MENU_IMAGES_FOR_INTERNAL_CLIENTS)

@interface NSMenuItem (Staging_138651669)
+ (NSString *)_systemImageNameForAction:(SEL)action;
@property (strong, setter=_setActionImage:) NSImage *_actionImage;
@end

#endif
```

whsieh commented on Source/WebCore/PAL/pal/spi/mac/NSMenuSPI.h:
```diff
@@ -79,6 +79,16 @@ enum {
 @end
 
+#if ENABLE(CONTEXT_MENU_IMAGES_FOR_INTERNAL_CLIENTS)
+@interface NSMenuItem (Staging_139537720)
```
Nit - rdar://139537720 seems to be the radar for this WebKit change. I think you meant rdar://138651669?

```suggestion
@interface NSMenuItem (Staging_138651669)
```

whsieh commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3029,7 +3029,18 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
-    if (auto *menuItem = singleMenuItemOrExtensionItemWithSubmenu(contextParameters))
+#if USE(APPKIT)
+#if ENABLE(CONTEXT_MENU_IMAGES_FOR_INTERNAL_CLIENTS)
+    WebPageProxy* page = [tab.webView() _page];
```
```suggestion
    RefPtr page = [tab.webView() _page];
```

whsieh commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3029,7 +3029,18 @@ static WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+#if USE(APPKIT)
+#if ENABLE(CONTEXT_MENU_IMAGES_FOR_INTERNAL_CLIENTS)
+    WebPageProxy* page = [tab.webView() _page];
+    bool allowTopLevelImages = page ? page->preferences().contextMenuImagesForInternalClientsEnabled() : false;
```
Nit - we generally do this:

```suggestion
    bool allowTopLevelImages = page && page->preferences().contextMenuImagesForInternalClientsEnabled();
```

whsieh commented on Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
```diff
@@ -884,9 +934,15 @@ - (void)menuDidClose:(NSMenu *)menu
+    case WebCore::ContextMenuItemType::CheckableAction: {
+        RetainPtr menuItem = createMenuActionItem(item);
+#if ENABLE(CONTEXT_MENU_IMAGES_FOR_INTERNAL_CLIENTS)
+        if (this->page()->preferences().contextMenuImagesForInternalClientsEnabled())
```
Nit - is the `this->` required?

aprotyas reviewed (approved):
Thanks for iterating!

lilyspiniolas committed:
Internal clients unable to set context menu images
https://bugs.webkit.org/show_bug.cgi?id=285317
rdar://139537720

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

Allow internal WebKit clients to enable context menu images.
When enabled, the ContextMenuAction of menu items is used
to determine which image to use, if available.

PDFContextMenuItems now store a ContextMenuAction so that
the image can be determined in the same manner. Each item's
ContextMenuAction is determined from its ContextMenuItemTag.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/PAL/pal/spi/mac/NSMenuSPI.h:
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKit/Platform/mac/MenuUtilities.h:
* Source/WebKit/Platform/mac/MenuUtilities.mm:
(WebKit::symbolNameForAction):
* Source/WebKit/Shared/mac/PDFContextMenu.h:
* Source/WebKit/Shared/mac/PDFContextMenuItem.serialization.in:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::platformMenuItems const):
(WebKit::WebExtensionContext::singleMenuItemOrExtensionItemWithSubmenu const):
(WebKit::WebExtensionContext::addItemsToContextMenu):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::updateMenuItemImage):
(WebKit::WebContextMenuProxyMac::getContextMenuItem):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::showPDFContextMenu):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::handleContextMenuEvent):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::contextMenuActionFromTag const):
(WebKit::UnifiedPDFPlugin::contextMenuItem const):
(WebKit::UnifiedPDFPlugin::separatorContextMenuItem const):

Canonical link: https://commits.webkit.org/288471@main
